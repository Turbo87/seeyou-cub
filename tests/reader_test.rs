use seeyou_cub::{CubReader, CubStyle};

#[test]
fn parse_france_fixture() {
    let mut reader = CubReader::from_path("tests/fixtures/france_2024.07.02.cub")
        .expect("Failed to open fixture file");
    let mut warnings = Vec::new();

    let header = reader
        .read_header(&mut warnings)
        .expect("Failed to parse header");

    // Validate warnings
    assert_eq!(warnings.len(), 0);

    // Snapshot the header
    insta::assert_debug_snapshot!("header", header);

    // Parse all items
    let items: Vec<_> = reader
        .read_items(&header, &mut warnings)
        .expect("Failed to create item iterator")
        .collect::<Result<Vec<_>, _>>()
        .expect("Failed to parse items");

    // Validate item count
    assert_eq!(items.len(), 1368);

    // Snapshot first few items
    insta::assert_debug_snapshot!("items_sample", &items[0..5]);

    // Validate first item's points
    let first_item = items.first().unwrap();
    let points: Vec<_> = reader
        .read_points(&header, first_item, &mut warnings)
        .expect("Failed to read points")
        .collect::<Result<Vec<_>, _>>()
        .expect("Failed to parse points");

    insta::assert_debug_snapshot!("first_item_points", points);

    let last_item = items.last().unwrap();
    let points: Vec<_> = reader
        .read_points(&header, last_item, &mut warnings)
        .expect("Failed to read points")
        .collect::<Result<Vec<_>, _>>()
        .expect("Failed to parse points");

    insta::assert_debug_snapshot!("last_item_points", points);

    // Validate all airspaces and collect statistics
    let mut total_points = 0;
    let mut style_counts = std::collections::HashMap::new();

    for item in &items {
        *style_counts.entry(item.style()).or_insert(0) += 1;

        let mut point_iter = reader
            .read_points(&header, item, &mut warnings)
            .expect("Failed to read points");

        for point_result in &mut point_iter {
            let _point = point_result.expect("Failed to parse point");
            total_points += 1;
        }
    }

    // Validate total points
    assert_eq!(total_points, 93784, "Total points mismatch");

    // Validate airspace type distribution
    assert_eq!(
        *style_counts.get(&CubStyle::Unknown).unwrap_or(&0),
        452,
        "Unknown airspace count mismatch"
    );
    assert_eq!(
        *style_counts.get(&CubStyle::RestrictedArea).unwrap_or(&0),
        435,
        "Restricted area count mismatch"
    );
    assert_eq!(
        *style_counts.get(&CubStyle::ProhibitedArea).unwrap_or(&0),
        113,
        "Prohibited area count mismatch"
    );
    assert_eq!(
        *style_counts.get(&CubStyle::DangerArea).unwrap_or(&0),
        111,
        "Danger area count mismatch"
    );
    assert_eq!(
        *style_counts.get(&CubStyle::GliderSector).unwrap_or(&0),
        135,
        "Glider sector count mismatch"
    );
    assert_eq!(
        *style_counts.get(&CubStyle::ControlZone).unwrap_or(&0),
        92,
        "Control zone count mismatch"
    );
}
