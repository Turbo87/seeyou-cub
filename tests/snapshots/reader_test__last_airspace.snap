---
source: tests/reader_test.rs
expression: last_airspace
---
Airspace {
    bounding_box: BoundingBox {
        left: 0.013671746,
        top: 0.87266463,
        right: 0.06952207,
        bottom: 0.83775806,
    },
    style: Unknown,
    class: ClassD,
    extended_type: None,
    min_alt: 3505,
    max_alt: 5944,
    min_alt_style: FlightLevel,
    max_alt_style: FlightLevel,
    time_out: 0,
    start_date: None,
    end_date: None,
    extra_data: 0,
    days_active: DaysActive {
        bits: 0,
    },
    points: [
        Point { lat: 0.8726647, lon: 0.01745355 },
        Point { lat: 0.8464845, lon: 0.01745355 },
        Point { lat: 0.8417673, lon: 0.013671746 },
        Point { lat: 0.8417673, lon: 0.023799524 },
        Point { lat: 0.8403761, lon: 0.02844422 },
        Point { lat: 0.8403761, lon: 0.03577932 },
        Point { lat: 0.83775806, lon: 0.043633223 },
        Point { lat: 0.83775806, lon: 0.06952207 },
        Point { lat: 0.8443853, lon: 0.06635653 },
        Point { lat: 0.8476627, lon: 0.06477084 },
        Point { lat: 0.85579306, lon: 0.060795687 },
        Point { lat: 0.86073816, lon: 0.058870696 },
        Point { lat: 0.8624932, lon: 0.05820648 },
        Point { lat: 0.8668471, lon: 0.056485493 },
        Point { lat: 0.87052184, lon: 0.055161864 },
        Point { lat: 0.8726647, lon: 0.054395914 },
        Point { lat: 0.8726647, lon: 0.017453548 },
    ],
    name: Some(
        "Paris1",
    ),
    frequency_name: None,
    icao_code: None,
    exception_rules: None,
    notam_remarks: None,
    notam_id: None,
    frequency: None,
    secondary_frequency: None,
    notam_insert_time: None,
}
